Add client-certificate based authentication for CockroachDB Cloud (#16779)
cockroachdb/docs: cockroachcloud/client-certs-dedicated.md

---

> title: Manage Certificate Authentication for SQL Clients in CockroachDB Dedicated Clusters

Author: I think for search page visibility/ SEO it's much better to have CockroachDB in the title. we don't per character.
This is a general style team issue I would say, I will raise it with @taroface @ianjevans @mikeCRL at style team meeting on Thursday.

> title: Managing Certificate Authentication for SQL Clients in CockroachDB Dedicated Clusters

Reviewer: ```suggestion
title: Manage Certificate Authentication for SQL Clients in CockroachDB Dedicated Clusters
```

> summary: procedures for managing client certificates for dedicated clusters

Reviewer: There is a Vale test that fails if you have capitalized "Serverless". There isn't the same Vale test for Dedicated (I don't know why). I think the second edit was because I don't think "dedicated cluster" looks right. I guess we are in a state of ambiguity and I bow to your preferences.

Refer to [Authenticating to {{ site.data.products.db }}](authentication.html) for an overview of authentication in {{ site.data.products.db }}, both at the level of the organization and at the callout

{{site.data.alerts.callout_info}}
This feature is in [**limited access**](../{{site.versions["stable"]}}/cockroachdb-feature-availability.html), and is only available to organizations that choose to opt-in. To enroll your organization, contact your Cockroach Labs account team. These features are subject to change.
{{site.data.alerts.end}}

## Provision your cluster's PKI hierarchy

There are many ways to create, manage and distribute digital security certificates. Cockroach Labs recommends using a secure secrets server such as [HashiCorp Vault](https://www.vaultproject.io/), which can be used to securely generate certificates, without the need to reveal the CA private key.

Refer to: [CockroachDB - HashiCorp Vault Integration](../{{site.versions["stable"]}}/hashicorp-integration.html)

> Alternatively, you can generate certificates using CockroachDB's `cockroach cert` command or with [OpenSSL](https://www.openssl.org/). However, generating certificates this way and manually handling cryptographic material comes with considerable additional risk and room for error. PKI cryptographic material related to your {{ site.data.products.db }} organizations, particularly in any production systems, should be handled according to a considered policy appropriate to your security goals.

Author: I prefer my way of putting it. What don't you like about it?

> Alternatively, you can generate certificates using CockroachDB's `cockroach cert` command or with [OpenSSL](https://www.openssl.org/). However, generating certificates this way and manually handling cryptographic material comes with considerable additional risk and room for error. PKI cryptographic material related to your {{ site.data.products.db }} organizations, particularly in any production systems, should be handled according to a considered policy appropriate to your security goals.

Reviewer: I guess your way doesn't seemed very indirect to me, we are distancing ourselves two levels: Your policy, your goals, but handle it correctly. I thought telling them to handle it as sensitive information gives them more info about what to do. Completely your call.

    1. Option 3: If using a Vault deployment internal to your organization, contact your Vault administrator for a token and the appropriate endpoint.

1. Initialize your shell for Vault.

    {% include_cached copy-clipboard.html %}
    ~~~shell
    export VAULT_ADDR= # your Vault cluster's Public URL
    export VAULT_NAMESPACE="admin"
    ~~~

1. Authenticate with your admin token 

    {% include_cached copy-clipboard.html %}
    ~~~shell
    vault login
    ~~~

### Create the certificate authority (CA) certificate

This CA certificate will be used to [configure your cluster's Trust Store](#upload-a-certificate-authority-ca-certificate-for-a-cockroachdb-dedicated-cluster). Any client certificate signed by the CA identified by this certificate will be trusted, and be able to authenticate to your cluster.

1. Create a PKI secrets engine to serve as your client CA.

    {% include_cached copy-clipboard.html %}
    ~~~shell
    vault secrets enable -path=cockroach_client_ca pki
    ~~~

    ~~~txt
    Success! Enabled the pki secrets engine at: cockroach_client_ca/
    ~~~
    

> 1. Generate a root credential pair for the CA. Certificates created with this CA/secrets engine will be signed with the private key generated here and held within Vault; this key cannot be exported, safeguarding it from being leaked and used to issue fraudulent certificates. The CA public certificate is downloaded in the resulting JSON payload.

Author: I prefer mine, is this a strong opinion?

> 1. Generate a root credential pair for the CA. Certificates created with this CA/secrets engine will be signed with the private key generated here and held within Vault; this key cannot be exported, safeguarding it from being leaked and used to issue fraudulent certificates. The CA public certificate is downloaded in the resulting JSON payload.

Reviewer: At the very least, I think the guarantee needs to be weakened. We can't guarantee that the CA won't be leaked just because they use Vault. That's what "help" is doing in my suggestion. Also, "safeguarding" seems weird and jargony as a verb.

> 1. Generate a root credential pair for the CA. Certificates created with this CA/secrets engine will be signed with the private key generated here and held within Vault; this key cannot be exported, safeguarding it from being leaked and used to issue fraudulent certificates. The CA public certificate is downloaded in the resulting JSON payload.

Author: "Safeguarding" is not jargon, it's regular english and not particularly obscure. Our readers can look an unfamiliar word up in the dictionary if they need to. I think the definitions offered by oxford or merriam webster are perfectly appropriate to what I mean. They do not imply a guarantee, just a measure intended for protection or precaution.

It's Vault's guarantee, not ours, that the key can't be exported:
https://developer.hashicorp.com/vault/docs/secrets/pki/setup
"The private key is safely stored internally in Vault."

> 1. Generate a root credential pair for the CA. Certificates created with this CA/secrets engine will be signed with the private key generated here and held within Vault; this key cannot be exported, safeguarding it from being leaked and used to issue fraudulent certificates. The CA public certificate is downloaded in the resulting JSON payload.

Reviewer: SGTM

    {% include_cached copy-clipboard.html %}
    ~~~shell
    vault write \
    cockroach_client_ca/root/generate/internal \
    ttl=1000h \
    --format=json > "${SECRETS_DIR}/certs/cockroach_client_ca_cert.json"
    ~~~

1. Format the public certificate for upload.

    In order to upload your certificate to {{ site.data.products.db }}, you must create a JSON file with the public certificate itself as the value for a key `x509_pem_cert`, as in the following example.

    ~~~json
    {
      "x509_pem_cert": "-----BEGIN CERTIFICATE-----\nMIIDfzCCAmagAwIBAgIBADANBgkqhkiG9w0BAQ0FADBZMQswCQYDVQQGEwJ1czEL\nMAkGA1UECAwCV0ExDTALBgNVBAoMBHRlc3QxDTALBgNVBAMMBHRlc3QxEDAOBgNV\nBAcMB1NlYXR0bGUxDTALBgNVBAsMBHRlc3QwHhcNMjMwMzE2MjMyNTMxWhcNMjQw\nMzE1MjMyNTMxWjBZMQswCQYDVQQGEwJ1czELMAkGA1UECAwCV0ExDTALBgNVBAoM\nBHRlc3QxDTALBgNVBAMMBHRlc3QxEDAOBgNVBAcMB1NlYXR0bGUxDTALBgNVBAsM...\n-----END CERTIFICATE-----"
    }
    ~~~

>     The public certificate can be found in the JSON file created by Vault at `.data.certificate`, for example, using the `jq` utility:

Author: yup, will do.

    ~~~shell
    echo -e $(cat "${SECRETS_DIR}/clients/certs.json" | jq .data.private_key | tr -d '"') > "${SECRETS_DIR}/clients/client.root.key"
    echo -e $(cat "${SECRETS_DIR}/clients/certs.json" | jq .data.certificate | tr -d '"') > "${SECRETS_DIR}/clients/client.root.crt"
    ~~~

1. Set key file permissions; CockroachDB will reject improperly permissioned private keys for authentication.

    {% include_cached copy-clipboard.html %}
    ~~~shell
    chmod 0600  ${SECRETS_DIR}/clients/client.root.key
    chown $USER ${SECRETS_DIR}/clients/client.root.key
    ~~~

## Upload a certificate authority (CA) certificate for a {{ site.data.products.dedicated }} cluster

Add a CA certificate to your cluster's trust store for client authentication. Client certificates signed using the private key corresponding to this certificate will be accepted by your cluster for certificate-based client authentication.

Refer to [Transport Layer Security (TLS) and Public Key Infrastructure (PKI): The CockroachDB certificate Trust Store](../{{site.versions["stable"]}}/security-reference/transport-layer-security.html#the-cockroachdb-certificate-trust-store)

{{site.data.alerts.callout_success}}

> Managing the certificate authority (CA) certificate for a {{ site.data.products.dedicated }} cluster requires the [Cluster Administrator](authorization.html#cluster-administrator) or [Org Administrator (legacy)](authorization.html#org-administrator-legacy) Organization role.

Reviewer: ```suggestion
The [Cluster Administrator](authorization.html#cluster-administrator) or [Org Administrator (legacy)](authorization.html#org-administrator-legacy) Organization role is required to manage the CA certificate for a {{ site.data.products.dedicated }} cluster.
```
After you explode CA you probably don't need to explode it again in the same topic.

{{site.data.alerts.end}}

<div class="filters clearfix">
  <button class="filter-button page-level" data-scope="api">Using the API</button>
  <button class="filter-button page-level" data-scope="tf">Using Terraform</button>
</div>
<p></p>

<section class="filter-content" markdown="1" data-scope="api">

> 1. Submit the asynchronous request, supplying your cluster ID, API key, and the path to the certificate JSON with your CA certificate, as described in [Create the certificate authority (CA) certificate](#create-the-certificate-authority-ca-certificate).

Reviewer: Indent the code block if it belongs with the step. (similar below)

It could be confusing to show the 200 response for an asynchronous request. Is it worth explaining that the 200 only indicates the success of the submission, not the success of the request?

{% include_cached copy-clipboard.html %}
~~~shell
curl --request POST \
  --url ${COCKROACH_SERVER}/api/v1/clusters/${CLUSTER_ID}/client-ca-cert \
  --header "Authorization: Bearer ${API_KEY}" \
  --header 'content-type: application/json' \
  --data "@cockroach_client_ca_cert.json"
~~~

~~~txt
200 OK
~~~

1. Confirm success of the operation.

{% include_cached copy-clipboard.html %}
~~~shell
curl --request GET \
  --url ${COCKROACH_SERVER}/api/v1/clusters/${CLUSTER_ID}/client-ca-cert \
  --header "Authorization: Bearer ${API_KEY}"
~~~

~~~txt
{
  "status": "PENDING",
  "x509_pem_cert": ""
}
~~~

~~~txt
{
  "status": "IS_SET",
  "x509_pem_cert": "-----BEGIN CERTIFICATE-----\nMIIDfzCCAmagAwIBAgIBADANBgkqhkiG9w0BAQ0FADBZMQswCQYDVQQGEwJ1czEL\nMAkGA1UECAwCV0ExDTALBgNVBAoMBHRlc3QxDTALBgNVBAMMBHRlc3QxEDAOBgNV\nBAcMB1NlYXR0bGUxDTALBgNVBAsMBHRlc3QwHhcNMjMwMzE2MjMyNTMxWhcNMjQw\n
...\n-----END CERTIFICATE-----",
}
~~~

</section>
<section class="filter-content" markdown="1" data-scope="tf">

> Add the following block to your terraform configuration and apply:

Reviewer: ```suggestion
Add the following block to your Terraform template and apply the change:
```

{% include_cached copy-clipboard.html %}
~~~shell
resource "cockroach_client_ca_cert" "yourclustername" {
  id = cockroach_cluster.example.id
  x509_pem_cert = file("cockroach_client_ca_cert.json")
}
~~~
</section>

> ## Update the certificate authority (CA) certificate for a dedicated cluster

Reviewer: I don't think you need "certificate authority" in these titles.
```suggestion
## Update the CA certificate for a dedicated cluster
```

> Replace the CA certificate used by your cluster for certificate-based client authentication.

Reviewer: ```suggestion
This section shows how to replace the CA certificate used by your cluster for certificate-based client authentication.
```

{{site.data.alerts.callout_success}}

> Managing the certificate authority (CA) certificate for a {{ site.data.products.dedicated }} cluster requires the [Cluster Administrator](authorization.html#cluster-administrator) or [Org Administrator (legacy)](authorization.html#org-administrator-legacy) Organization role.

Reviewer: If you take the suggestion above for similar text, please sync it here.

{{site.data.alerts.end}}

<div class="filters clearfix">
  <button class="filter-button page-level" data-scope="api">Using the API</button>
  <button class="filter-button page-level" data-scope="tf">Using Terraform</button>
</div>
<p></p>

<section class="filter-content" markdown="1" data-scope="api">

{% include_cached copy-clipboard.html %}
~~~shell
curl --request PATCH \
  --url ${COCKROACH_SERVER}/api/v1/clusters/${CLUSTER_ID}/client-ca-cert \
  --header "Authorization: Bearer ${API_KEY}" \
  --header 'content-type: application/json' \
  --data "@cockroach_client_ca_cert.json"
~~~

~~~txt
200 OK
~~~
</section>

<section class="filter-content" markdown="1" data-scope="tf">

> Update the `cockroach_client_ca_cert` block in your terraform configuration, pointing to the updated certificate JSON file, and apply:

Reviewer: ```suggestion
Update the `cockroach_client_ca_cert` block in your Terraform template to point to a JSON file that contains the new certificate, and apply the change:
```

{% include_cached copy-clipboard.html %}
~~~shell
resource "cockroach_client_ca_cert" "yourclustername" {
  id = cockroach_cluster.example.id
  x509_pem_cert = file("cockroach_client_ca_cert.json")
}
~~~
</section>

## Delete the certificate authority (CA) certificate for a dedicated cluster

> Remove the configured CA certificate from the cluster. Clients will no longer be able to authenticate with certificates signed by this CA certificate.

Reviewer: ```suggestion
This section shows how to remove the configured CA certificate from the cluster. Afterward, clients can no longer authenticate with certificates signed by this CA certificate.
```

{{site.data.alerts.callout_success}}

> Managing the certificate authority (CA) certificate for a {{ site.data.products.dedicated }} cluster requires the [Cluster Administrator](authorization.html#cluster-administrator) or [Org Administrator (legacy)](authorization.html#org-administrator-legacy) Organization role.

Reviewer: Ditto for keeping this in sync

Author: sorry Matt, I missed this comment and now am not sure how to find which previous , comment you were referring to, do you recall the suggestion?

{{site.data.alerts.end}}

<div class="filters clearfix">
  <button class="filter-button page-level" data-scope="api">Using the API</button>
  <button class="filter-button page-level" data-scope="tf">Using Terraform</button>
</div>
<p></p>

<section class="filter-content" markdown="1" data-scope="api">

{% include_cached copy-clipboard.html %}
~~~shell
curl --request DELETE \
  --url ${COCKROACH_SERVER}/api/v1/clusters/${CLUSTER_ID}/client-ca-cert \
  --header "Authorization: Bearer ${API_KEY}"
~~~

~~~txt
200 OK
~~~
</section>
<section class="filter-content" markdown="1" data-scope="tf">

To delete the client CA cert on a cluster, remove the `cockroach_client_ca_cert` resource from your terraform configuration, then run `terraform apply`.
</section>

## Authenticate a SQL client to a {{ site.data.products.dedicated }} cluster

> To use certificate authentication for a SQL client, you must include the filepaths to the client's private key and public certificate. The public certificate must be signed by a certificate authority (CA) who's public certificate is configured to be trusted by the cluster. Refer to [Upload a certificate authority (CA) certificate for a {{ site.data.products.dedicated }} cluster](#upload-a-certificate-authority-ca-certificate-for-a-cockroachdb-dedicated-cluster)

Reviewer: ```suggestion
To use certificate authentication for a SQL client, you must include the filepaths to the client's private key and public certificate. The public certificate must be signed by a CA that the cluster has been configured to trust. Refer to [Upload a certificate authority (CA) certificate for a {{ site.data.products.dedicated }} cluster](#upload-a-certificate-authority-ca-certificate-for-a-cockroachdb-dedicated-cluster).
```

> 1. From your cluster's overview page, `https://cockroachlabs.cloud/cluster/{ your cluster ID }`, click on the **Connect** button.

Reviewer: ```suggestion
1. From your cluster's overview page, `https://cockroachlabs.cloud/cluster/{ your cluster ID }`, click the **Connect** button.
```

> 1. Execute the command listed under **Download CA Cert** to download the required public certificate. This is used by your client to verify the identity of the cluster.

Reviewer: ```suggestion
1. Copy the command listed under **Download CA Cert** and run it locally to download the required public certificate, which your client will use to verify the identity of the cluster.
```

> 1. Obtain a connection string/command for your cluster. This connection string is designed for password authentication and must be modified.

Reviewer: Is this also available in the UI or how do you get it?

> 1. Obtain a connection string/command for your cluster. This connection string is designed for password authentication and must be modified.

Author: ```suggestion
1. Obtain your choice of connection string or CLI connection command for your cluster from the UI. This connection string is designed for password authentication and must be modified.
```

    1. Remove the placeholder password from the connection string.

>     1. Construct the full connection string by providing the paths to `sslrootcert` (the cluster's public CA certificate), `sslcert` (the client's public certificate, which must be signed by the same CA), and `sslkey` (the client's private key).

Reviewer: Is it worth giving an example here of what the connection string will look like?

Author: There are a lot of different types and they are changing in the UI, so the vagueness is a choice of necessity.

>     1. Construct the full connection string by providing the paths to `sslrootcert` (the cluster's public CA certificate), `sslcert` (the client's public certificate, which must be signed by the same CA), and `sslkey` (the client's private key).

Reviewer 2: >`sslcert` (the client's public certificate, which must be signed by the same CA)

Maybe be explicit here and say "signed by the client's CA".  "Same CA" may sound like we're talking about the cluster CA?

>     1. Construct the full connection string by providing the paths to `sslrootcert` (the cluster's public CA certificate), `sslcert` (the client's public certificate, which must be signed by the same CA), and `sslkey` (the client's private key).

Author: ```suggestion
    1. Construct the full connection string by providing the paths to `sslrootcert` (the cluster's public CA certificate), `sslcert` (the client's public certificate, which must be signed by the CA specified in `sslrootcert`), and `sslkey` (the client's private key).
```

---
title: Manage Certificate Authentication for SQL Clients
summary: procedures for managing client certificates for dedicated clusters
toc: true
docs_area: manage.security
---

> SQL clients may authenticate to {{ site.data.products.dedicated }} clusters using public key infrastructure (PKI) security certificates as an alternative to authenticating using a username and password or using [Cluster Single Sign-on (SSO) using CockroachDB Cloud Console](cloud-sso-sql.html) or [Cluster Single Sign-on (SSO) using JSON web tokens (JWT)](../{{site.versions["stable"]}}/sso-sql.html). This page describes how to administer the cluster's certificate authority (CA) certificate and authenticate to a cluster using client certificates.

Reviewer 3: SSO using CC Console / CLI is still suggested, so we should present it such that client certs are complimentary to that. That way, customers could use SSO for human users logging in for ad-hoc operations via CLI, and apps could connect via certs (as an alternative to JWT).

toc: true
docs_area: manage.security
cloud: true
---

SQL clients may authenticate to {{ site.data.products.dedicated }} clusters using public key infrastructure (PKI) security certificates as an alternative to authenticating using a username and password or using [Cluster Single Sign-on (SSO) using CockroachDB Cloud Console](cloud-sso-sql.html) or [Cluster Single Sign-on (SSO) using JSON web tokens (JWT)](../{{site.versions["stable"]}}/sso-sql.html). This page describes how to administer the cluster's certificate authority (CA) certificate and authenticate to a cluster using client certificates.

Refer to [Transport Layer Security (TLS) and Public Key Infrastructure (PKI)](../{{site.versions["stable"]}}/security-reference/transport-layer-security.html) for an overview of PKI certificate authentication in general and its use in CockroachDB.

Refer to [Authenticating to {{ site.data.products.db }}](authentication.html) for an overview of authentication in {{ site.data.products.db }}, both at the level of the organization and at the cluster.

{{site.data.alerts.callout_info}}
{% include_cached feature-phases/limited-access.md %}
{{site.data.alerts.end}}

> ## Provision your cluster's PKI hierarchy

Reviewer 3: Should we name this something like "Provision PKI hierarchy for SQL authentication in your cluster"? Basically I want to avoid giving an indication that node-level certs could also be managed in cloud.

There are many ways to create, manage, and distribute digital security certificates. Cockroach Labs recommends using a secure secrets server such as [HashiCorp Vault](https://www.vaultproject.io/), which can be used to securely generate certificates without revealing the CA private key.

Refer to: [CockroachDB - HashiCorp Vault Integration](../{{site.versions["stable"]}}/hashicorp-integration.html)

Alternatively, you can generate certificates using CockroachDB's `cockroach cert` command or with [OpenSSL](https://www.openssl.org/). However, generating certificates this way and manually handling cryptographic material comes with considerable additional risk and room for error. PKI cryptographic material related to your {{ site.data.products.db }} organizations, particularly in any production systems, should be handled according to a considered policy appropriate to your security goals.

> Refer to:

Reviewer 3: Again, I think we should add a note here that one should refer those pages only in the context of SQL client certs and not the node certs.

Author: made this a little sleeker.

      --header "Authorization: Bearer ${API_KEY}" \
      --header 'content-type: application/json' \
      --data "@cockroach_client_ca_cert.json"
    ~~~

    ~~~txt
    200 OK
    ~~~

1. Confirm success of the operation with the following `GET` request.

    {% include_cached copy-clipboard.html %}
    ~~~shell
    curl --request GET \
      --url ${COCKROACH_SERVER}/api/v1/clusters/${CLUSTER_ID}/client-ca-cert \
      --header "Authorization: Bearer ${API_KEY}"
    ~~~

    `PENDING` indicates that the operation is still in process.
    
    ~~~txt
    {
      "status": "PENDING",
      "x509_pem_cert": ""
    }
    ~~~
    
    `IS_SET` indicates that the operation completed successfully, confirming the configured public CA cert.

    ~~~txt
    {
      "status": "IS_SET",
      "x509_pem_cert": "-----BEGIN CERTIFICATE-----\nMIIDfzCCAmagAwIBAgIBADANBgkqhkiG9w0BAQ0FADBZMQswCQYDVQQGEwJ1czEL\nMAkGA1UECAwCV0ExDTALBgNVBAoMBHRlc3QxDTALBgNVBAMMBHRlc3QxEDAOBgNV\nBAcMB1NlYXR0bGUxDTALBgNVBAsMBHRlc3QwHhcNMjMwMzE2MjMyNTMxWhcNMjQw\n
    ...\n-----END CERTIFICATE-----",
    }
    ~~~

</section>
<section class="filter-content" markdown="1" data-scope="tf">

> Add the following block to your Terraform template and apply the change:

Reviewer 3: Maybe also link to the resource documentation here - https://registry.terraform.io/providers/cockroachdb/cockroach/latest/docs/resources/client_ca_cert?

Author: ok

{% include_cached copy-clipboard.html %}
~~~shell
resource "cockroach_client_ca_cert" "yourclustername" {
  id = cockroach_cluster.example.id
  x509_pem_cert = file("cockroach_client_ca_cert.json")
}
~~~
</section>

## Update the CA certificate for a dedicated cluster

> This section shows how to replace the CA certificate used by your cluster for certificate-based client authentication.

Reviewer 3: Can we add a recommendation here, that this action for CA cert upload should be done by informing all the existing users, as we issue SIGHUP to the cluster behind the scenes which may impact existing connections?

Author: How does this look?

Reviewer 3: LGTM, thank you

    A `200` successful response code indicates that the asynchronous request was successfully submitted, but does not guarantee that the operation (configuring the CA certificate) successfully completed. You must confirm success with a follow-up `GET` request, as described in the next step.
 
    {% include_cached copy-clipboard.html %}
    ~~~shell
    curl --request DELETE \
      --url ${COCKROACH_SERVER}/api/v1/clusters/${CLUSTER_ID}/client-ca-cert \
      --header "Authorization: Bearer ${API_KEY}"
    ~~~

    ~~~txt
    200 OK
    ~~~

1. Confirm success of the operation with the following `GET` request.

    {% include_cached copy-clipboard.html %}
    ~~~shell
    curl --request GET \
      --url ${COCKROACH_SERVER}/api/v1/clusters/${CLUSTER_ID}/client-ca-cert \
      --header "Authorization: Bearer ${API_KEY}"
    ~~~

    `PENDING` indicates that the operation is still in process.
    
    ~~~txt
    {
      "status": "PENDING",
      "x509_pem_cert": ""
    }
    ~~~
    

>     `IS_SET` indicates that the operation completed successfully, confirming the configured public CA cert.

Reviewer 3: This is the DELETE operation, so shouldn't be IS_SET, right? NOT_SET would be the correct response. Check this section - https://docs.google.com/document/d/1uyQuPmJ4NawuvaGTODr8AVArh7P-5TrdnWxELD6ppWI/edit#heading=h.47j5fsmw3vlu

Author: Thanks

    {
      "status": "IS_SET",
      "x509_pem_cert": "-----BEGIN CERTIFICATE-----\nMIIDfzCCAmagAwIBAgIBADANBgkqhkiG9w0BAQ0FADBZMQswCQYDVQQGEwJ1czEL\nMAkGA1UECAwCV0ExDTALBgNVBAoMBHRlc3QxDTALBgNVBAMMBHRlc3QxEDAOBgNV\nBAcMB1NlYXR0bGUxDTALBgNVBAsMBHRlc3QwHhcNMjMwMzE2MjMyNTMxWhcNMjQw\n
    ...\n-----END CERTIFICATE-----",
    }
    ~~~
</section>
<section class="filter-content" markdown="1" data-scope="tf">

To delete the client CA cert on a cluster, remove the `cockroach_client_ca_cert` resource from your Terraform template, then run `terraform apply`.
</section>

## Authenticate a SQL client to a {{ site.data.products.dedicated }} cluster

To use certificate authentication for a SQL client, you must include the filepaths to the client's private key and public certificate. The public certificate must be signed by a CA that the cluster has been configured to trust. Refer to [Upload a certificate authority (CA) certificate for a {{ site.data.products.dedicated }} cluster](#upload-a-certificate-authority-ca-certificate-for-a-cockroachdb-dedicated-cluster).

1. From your cluster's overview page, `https://cockroachlabs.cloud/cluster/{ your cluster ID }`, click the **Connect** button.

> 1. Copy the command listed under **Download CA Cert** and run it locally to download the required public certificate, which your client will use to verify the identity of the cluster.

Reviewer 3: Maybe clarify here that this CA certificate is for the cluster, to prove the server identity to the SQL client. It shouldn't be confused with the CA cert created and uploaded above.

Author: It says "which your client will use to verify the identity of the cluster", not sure what else to say?

Reviewer 3: Alright, we can leave it as is.
